PersistenceTest.when_leaderIsRestarted_then_itRestoresItsRaftStateWithSnapshotAndBecomesLeader

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3647
```
java.lang.AssertionError: expected same:<RaftNode{groupId=TestRaftGroupId{name='test''}, status=ACTIVE}> was not:<RaftNode{groupId=TestRaftGroupId{name='test''}, status=ACTIVE}>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotSame(Assert.java:828)
	at org.junit.Assert.assertSame(Assert.java:771)
	at org.junit.Assert.assertSame(Assert.java:782)
	at com.hazelcast.cp.internal.raft.impl.PersistenceTest.when_leaderIsRestarted_then_itRestoresItsRaftStateWithSnapshotAndBecomesLeader(PersistenceTest.java:512)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

PersistenceTest.when_followerIsRestarted_then_itRestoresItsRaftState

http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-CorrettoJDK8/33/testReport/junit/com.hazelcast.cp.internal.raft.impl/PersistenceTest/when_followerIsRestarted_then_itRestoresItsRaftState/

**Stacktrace**
```
java.lang.AssertionError: expected:<10> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cp.internal.raft.impl.PersistenceTest.when_followerIsRestarted_then_itRestoresItsRaftState(PersistenceTest.java:365)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/RaftNodeImpl.java
@@ -1434,6 +1434,6 @@ public final class RaftNodeImpl implements RaftNode {
 
     @Override
     public String toString() {
-        return "RaftNode{" + "groupId=" + groupId + ", status=" + status + '}';
+        return "RaftNode{" + "groupId=" + groupId + ", status=" + status + ", localEndpoint=" + state.localEndpoint() + '}';
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/handler/VoteRequestHandlerTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/handler/VoteRequestHandlerTask.java
@@ -60,8 +60,11 @@ public class VoteRequestHandlerTask extends RaftNodeStatusAwareTask implements R
         // (Raft thesis - Section 4.2.3) This check conflicts with the leadership transfer mechanism,
         // in which a server legitimately starts an election without waiting an election timeout.
         // Those VoteRequest objects are marked with a special flag ("disruptive") to bypass leader stickiness.
-        if (!req.isDisruptive() && raftNode.lastAppendEntriesTimestamp()
-                > Clock.currentTimeMillis() - raftNode.getLeaderElectionTimeoutInMillis()) {
+        // Also if request comes from the current leader, then stickiness check is skipped.
+        // Since current leader may have restarted by recovering its persistent state.
+        long leaderElectionTimeoutDeadline = Clock.currentTimeMillis() - raftNode.getLeaderElectionTimeoutInMillis();
+        if (!req.isDisruptive() && raftNode.lastAppendEntriesTimestamp() > leaderElectionTimeoutDeadline
+                && !req.candidate().equals(state.leader())) {
             logger.info("Rejecting " + req + " since received append entries recently.");
             raftNode.send(new VoteResponse(localMember, state.term(), false), req.candidate());
             return;

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/PersistenceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/PersistenceTest.java
@@ -20,6 +20,7 @@ import com.hazelcast.config.cp.RaftAlgorithmConfig;
 import com.hazelcast.cp.internal.raft.impl.dataservice.ApplyRaftRunnable;
 import com.hazelcast.cp.internal.raft.impl.dataservice.RaftDataService;
 import com.hazelcast.cp.internal.raft.impl.dto.AppendRequest;
+import com.hazelcast.cp.internal.raft.impl.dto.PreVoteRequest;
 import com.hazelcast.cp.internal.raft.impl.log.LogEntry;
 import com.hazelcast.cp.internal.raft.impl.log.SnapshotEntry;
 import com.hazelcast.cp.internal.raft.impl.persistence.RaftStateStore;
@@ -315,8 +316,7 @@ public class PersistenceTest extends HazelcastTestSupport {
 
     @Test
     public void when_leaderIsRestarted_then_itRestoresItsRaftStateAndBecomesLeader() throws ExecutionException, InterruptedException {
-        RaftAlgorithmConfig config = new RaftAlgorithmConfig().setLeaderHeartbeatPeriodInMillis(SECONDS.toMillis(30));
-        group = new LocalRaftGroupBuilder(3, config).setRaftStateStoreFactory(RAFT_STATE_STORE_FACTORY)
+        group = new LocalRaftGroupBuilder(3).setRaftStateStoreFactory(RAFT_STATE_STORE_FACTORY)
                                             .setAppendNopEntryOnLeaderElection(true)
                                             .build();
         group.start();
@@ -334,8 +334,12 @@ public class PersistenceTest extends HazelcastTestSupport {
         InMemoryRaftStateStore stateStore = getRaftStateStore(leader);
         RestoredRaftState terminatedState = stateStore.toRestoredRaftState();
 
+        // Block voting between followers
+        // to avoid a leader election before leader restarts.
+        blockVotingBetweenFollowers();
+
         group.terminateNode(terminatedEndpoint);
-        final RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
+        RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
 
         RaftNodeImpl newLeader = group.waitUntilLeaderElected();
         assertSame(newLeader, restartedNode);
@@ -362,7 +366,7 @@ public class PersistenceTest extends HazelcastTestSupport {
             leader.replicate(new ApplyRaftRunnable("val" + i)).get();
         }
 
-        assertEquals(getCommitIndex(leader), getCommitIndex(terminatedFollower));
+        assertTrueEventually(() -> assertEquals(getCommitIndex(leader), getCommitIndex(terminatedFollower)));
 
         RaftEndpoint terminatedEndpoint = terminatedFollower.getLocalMember();
         InMemoryRaftStateStore stateStore = getRaftStateStore(terminatedFollower);
@@ -484,8 +488,8 @@ public class PersistenceTest extends HazelcastTestSupport {
     @Test
     public void when_leaderIsRestarted_then_itRestoresItsRaftStateWithSnapshotAndBecomesLeader() throws ExecutionException, InterruptedException {
         final int committedEntryCountToSnapshot = 50;
-        RaftAlgorithmConfig config = new RaftAlgorithmConfig().setCommitIndexAdvanceCountToSnapshot(committedEntryCountToSnapshot)
-                                                              .setLeaderHeartbeatPeriodInMillis(SECONDS.toMillis(30));
+        RaftAlgorithmConfig config = new RaftAlgorithmConfig()
+                .setCommitIndexAdvanceCountToSnapshot(committedEntryCountToSnapshot);
         group = new LocalRaftGroupBuilder(3, config).setAppendNopEntryOnLeaderElection(true)
                                                     .setRaftStateStoreFactory(RAFT_STATE_STORE_FACTORY)
                                                     .build();
@@ -505,15 +509,19 @@ public class PersistenceTest extends HazelcastTestSupport {
         InMemoryRaftStateStore stateStore = getRaftStateStore(leader);
         RestoredRaftState terminatedState = stateStore.toRestoredRaftState();
 
-        group.terminateNode(terminatedEndpoint);
-        final RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
+        // Block voting between followers
+        // to avoid a leader election before leader restarts.
+        blockVotingBetweenFollowers();
 
-        final RaftNodeImpl newLeader = group.waitUntilLeaderElected();
+        group.terminateNode(terminatedEndpoint);
+        RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
+
+        RaftNodeImpl newLeader = group.waitUntilLeaderElected();
         assertSame(restartedNode, newLeader);
 
         assertTrueEventually(() -> {
-            assertTrue(getTerm(newLeader) > term);
-            assertEquals(commitIndex + 1, getCommitIndex(newLeader));
+            assertTrue(getTerm(restartedNode) > term);
+            assertEquals(commitIndex + 1, getCommitIndex(restartedNode));
             RaftDataService service = group.getService(restartedNode);
             for (int i = 0; i <= committedEntryCountToSnapshot; i++) {
                 assertEquals("val" + i, service.get(i + 2));
@@ -521,10 +529,18 @@ public class PersistenceTest extends HazelcastTestSupport {
         });
     }
 
+    private void blockVotingBetweenFollowers() {
+        RaftEndpoint[] endpoints = group.getFollowerEndpoints();
+        for (RaftEndpoint endpoint : endpoints) {
+            if (group.isRunning(endpoint)) {
+                group.dropMessagesToAll(endpoint, PreVoteRequest.class);
+            }
+        }
+    }
+
     @Test
     public void when_leaderIsRestarted_then_itBecomesLeaderAndAppliesPreviouslyCommittedMemberList() throws ExecutionException, InterruptedException {
-        RaftAlgorithmConfig config = new RaftAlgorithmConfig().setLeaderHeartbeatPeriodInMillis(SECONDS.toMillis(30));
-        group = new LocalRaftGroupBuilder(3, config).setAppendNopEntryOnLeaderElection(true)
+        group = new LocalRaftGroupBuilder(3).setAppendNopEntryOnLeaderElection(true)
                                             .setRaftStateStoreFactory(RAFT_STATE_STORE_FACTORY)
                                             .build();
         group.start();
@@ -532,7 +548,7 @@ public class PersistenceTest extends HazelcastTestSupport {
         RaftNodeImpl leader = group.waitUntilLeaderElected();
         RaftNodeImpl[] followers = group.getNodesExcept(leader.getLocalMember());
         RaftNodeImpl removedFollower = followers[0];
-        final RaftNodeImpl runningFollower = followers[1];
+        RaftNodeImpl runningFollower = followers[1];
 
         group.terminateNode(removedFollower.getLocalMember());
         leader.replicate(new ApplyRaftRunnable("val")).get();
@@ -542,8 +558,12 @@ public class PersistenceTest extends HazelcastTestSupport {
         InMemoryRaftStateStore stateStore = getRaftStateStore(leader);
         RestoredRaftState terminatedState = stateStore.toRestoredRaftState();
 
+        // Block voting between followers
+        // to avoid a leader election before leader restarts.
+        blockVotingBetweenFollowers();
+
         group.terminateNode(terminatedEndpoint);
-        final RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
+        RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
 
         RaftNodeImpl newLeader = group.waitUntilLeaderElected();
         assertSame(restartedNode, newLeader);
@@ -594,8 +614,7 @@ public class PersistenceTest extends HazelcastTestSupport {
     @Test
     public void when_leaderIsRestarted_then_itBecomesLeaderAndAppliesPreviouslyCommittedMemberListViaSnapshot() throws ExecutionException, InterruptedException {
         int committedEntryCountToSnapshot = 50;
-        RaftAlgorithmConfig config = new RaftAlgorithmConfig().setCommitIndexAdvanceCountToSnapshot(committedEntryCountToSnapshot)
-                                                              .setLeaderHeartbeatPeriodInMillis(SECONDS.toMillis(30));
+        RaftAlgorithmConfig config = new RaftAlgorithmConfig().setCommitIndexAdvanceCountToSnapshot(committedEntryCountToSnapshot);
         group = new LocalRaftGroupBuilder(3, config).setAppendNopEntryOnLeaderElection(true)
                                                     .setRaftStateStoreFactory(RAFT_STATE_STORE_FACTORY)
                                                     .build();
@@ -618,8 +637,12 @@ public class PersistenceTest extends HazelcastTestSupport {
         InMemoryRaftStateStore stateStore = getRaftStateStore(leader);
         RestoredRaftState terminatedState = stateStore.toRestoredRaftState();
 
+        // Block voting between followers
+        // to avoid a leader election before leader restarts.
+        blockVotingBetweenFollowers();
+
         group.terminateNode(terminatedEndpoint);
-        final RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
+        RaftNodeImpl restartedNode = group.createNewRaftNode(terminatedState, stateStore);
 
         RaftNodeImpl newLeader = group.waitUntilLeaderElected();
         assertSame(restartedNode, newLeader);

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/testing/LocalRaftGroup.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/testing/LocalRaftGroup.java
@@ -330,6 +330,23 @@ public class LocalRaftGroup {
         return nodes[getIndexOfRunning(endpoint)];
     }
 
+    public RaftEndpoint[] getFollowerEndpoints() {
+        RaftEndpoint leaderEndpoint = getLeaderEndpoint();
+        RaftEndpoint[] n = new RaftEndpoint[members.length - 1];
+        int i = 0;
+        for (RaftEndpoint member : members) {
+            if (!member.equals(leaderEndpoint)) {
+                n[i++] = member;
+            }
+        }
+
+        if (i != n.length) {
+            throw new IllegalArgumentException();
+        }
+
+        return n;
+    }
+
     public RaftEndpoint getEndpoint(int index) {
         return members[index];
     }
@@ -639,5 +656,16 @@ public class LocalRaftGroup {
 
     public void terminateNode(RaftEndpoint endpoint) {
         terminateNode(getIndexOfRunning(endpoint));
+    }
+
+    public boolean isRunning(RaftEndpoint endpoint) {
+        Assert.assertNotNull(endpoint);
+        for (int i = 0; i < members.length; i++) {
+            if (!endpoint.equals(members[i])) {
+                continue;
+            }
+            return !integrations[i].isShutdown();
+        }
+        throw new IllegalArgumentException("Unknown endpoint: " + endpoint);
     }
 }

